### DISCUSSION TITLE
Failed to write Error: ENOENT. How to solve and understand this error?!

### DISCUSSION THREAD

**Discussion by J4v4Scr1pt** [vercel/next.js#56690] [Help]
### Summary

I have a nextjs@latest application that runs perfectly locally. And in production.
But on the nodeServer I see these errors from time to time.

`Failed to write image to cache 0g-agIzWCv1tkoUXVScWsNPsotuhEaMDFJat8K6khOg= [Error: ENOENT: no such file or directory, mkdir '/app/.next/cache'] {errno: -2,code: 'ENOENT',syscall: 'mkdir',path: '/app/.next/cache'}`

- How could I approach this issue? 
- And how can it still work?
- Will there be any performance whit this?

My thought is to copy the cache folder during the docker build, good idea?
![image](https://github.com/vercel/next.js/assets/5929055/13723390-cdd0-4c0d-bc2c-0d3c0afc6885)

```
FROM node:18-alpine AS base

# Step 1. Rebuild the source code only when needed
FROM base AS builder

WORKDIR /app

# Install dependencies based on the preferred package manager
COPY package.json yarn.lock* package-lock.json* pnpm-lock.yaml* ./
# Copy Authentication
COPY .npmrc .
# Omit --production flag for TypeScript devDependencies
RUN \
  if [ -f yarn.lock ]; then yarn --frozen-lockfile; \
  elif [ -f package-lock.json ]; then npm ci; \
  elif [ -f pnpm-lock.yaml ]; then yarn global add pnpm && pnpm i; \
  # Allow install without lockfile, so example works even without Node.js installed locally
  else echo "Warning: Lockfile not found. It is recommended to commit lockfiles to version control." && yarn install; \
  fi
COPY . .
  # if this is the sandbox deployment environment

# Environment variables must be present at build time
# https://github.com/vercel/next.js/discussions/14030

ARG NODE_ENV
ENV NODE_ENV=production
ARG NEXTAUTH_SECRET
ENV NEXTAUTH_SECRET=${SECRET}
ARG VERSION
ENV VERSION=${VERSION}
ARG SOURCEVERSION
ENV SOURCEVERSION=${SOURCEVERSION}

# Next.js collects completely anonymous telemetry data about general usage. Learn more here: https://nextjs.org/telemetry
# Uncomment the following line to disable telemetry at build time
# ENV NEXT_TELEMETRY_DISABLED 1

# Build Next.js based on the preferred package manager
RUN \
  if [ -f yarn.lock ]; then yarn build; \
  elif [ -f package-lock.json ]; then npm run build; \
  elif [ -f pnpm-lock.yaml ]; then pnpm build; \
  else yarn build; \
  fi

# Note: It is not necessary to add an intermediate step that does a full copy of `node_modules` here

# Step 2. Production image, copy all the files and run next
FROM base AS runner

WORKDIR /app

# Don't run production as root
RUN addgroup --system --gid 1001 nodejs
RUN adduser --system --uid 1001 nextjs
USER nextjs

COPY --from=builder /app/public ./public

# Automatically leverage output traces to reduce image size
# https://nextjs.org/docs/advanced-features/output-file-tracing
COPY --from=builder --chown=nextjs:nodejs /app/.next/standalone ./
COPY --from=builder --chown=nextjs:nodejs /app/.next/static ./.next/static

# Environment variables must be redefined at run time
ARG NODE_ENV
ENV NODE_ENV=production
ARG NEXTAUTH_SECRET
ENV NEXTAUTH_SECRET=${SECRET}
ARG VERSION
ENV VERSION=${VERSION}
ARG SOURCEVERSION
ENV SOURCEVERSION=${SOURCEVERSION}

# Uncomment the following line to disable telemetry at run time
# ENV NEXT_TELEMETRY_DISABLED 1

# Note: Don't expose ports here, Compose will handle that for us

CMD ["node", "server.js"]
```

### Additional information

```js
Operating System:
  Platform: win32
  Arch: x64
  Version: Windows 10 Pro
Binaries:
  Node: 20.7.0
  npm: N/A
  Yarn: N/A
  pnpm: 8.8.0
Relevant Packages:
  next: 13.5.4
  eslint-config-next: 13.5.4
  react: 18.2.0
  react-dom: 18.2.0
  typescript: 5.2.2
Next.js Config:
  output: standalone
```

**Comment by J4v4Scr1pt** [comment]
Found the solution to this matter:
I have checked all the examples with Docker when I built the docker-file. None of them had the solution expect one:
https://github.com/vercel/next.js/blob/canary/examples/with-docker/Dockerfile

I added this part:
```
# Set the correct permission for prerender cache
RUN mkdir .next
RUN chown nextjs:nodejs .next
``` 

**Edit:**
_I jumped to early.. this did **NOT** fix the issue._

**Comment by J4v4Scr1pt** [comment]
Maybe create an empty cache folder during build will fix it?

**Comment by ru4ert** [comment]
did you try
``` bash
RUN mkdir -p .next/cache 
RUN chown -R nextjs:nodejs .next
```
after `COPY --from=builder /app/public ./public`

**Comment by vanduc1102** [comment]
I am using `next@14.2.3` `nodejs@20`
I have the same issue, and have to use `root` user to by pass the issue. 
```

2024-05-18T15:54:24.921Z	⨯ Failed to write image to cache zmKE2pKz27GdFQ4bl4RJ7Iq8+-yUXRFqRDowaGfGGuY= Error: ENOENT: no such file or directory, mkdir '/app/apps/product-web/.next/cache'
2024-05-18T15:54:24.921Z	at async Object.mkdir (node:internal/fs/promises:855:10)
2024-05-18T15:54:24.921Z	at async writeToCacheDir (/app/node_modules/next/dist/server/image-optimizer.js:176:5)
2024-05-18T15:54:24.921Z	at async ImageOptimizerCache.set (/app/node_modules/next/dist/server/image-optimizer.js:416:13)
2024-05-18T15:54:24.921Z	at async /app/node_modules/next/dist/server/response-cache/index.js:121:25
2024-05-18T15:54:24.921Z	at async /app/node_modules/next/dist/lib/batcher.js:45:32 {
2024-05-18T15:54:24.921Z	errno: -2,
2024-05-18T15:54:24.921Z	code: 'ENOENT',
2024-05-18T15:54:24.921Z	syscall: 'mkdir',
2024-05-18T15:54:24.921Z	path: '/app/apps/product-web/.next/cache'
2024-05-18T15:54:24.921Z	}
```

RESOLVED

My application is running on AWS Fargate with `ReadonlyRootFilesystem:true` , which causing the error, changed to `ReadonlyRootFilesystem:false` to  resolve the issue.

Here is my `Dockerfile`:

```dockerfile
# turborepo where apps/product-web is an nextjs application
# production-web: NextJs application
FROM node:20-bookworm-slim AS web-runner

# Create a shop user to run the app
RUN addgroup --system --gid 1001 nodejs \
  && adduser --system --uid 1001 --home /home/shop --shell /bin/bash shop \
  && mkdir -p /app && chown -R shop:nodejs /app


WORKDIR /app

ARG ENVIRONMENT

ENV PORT 3000
ENV NODE_ENV=production

# Copy all needed for nextjs production
COPY  --chown=shop:nodejs --from=installer /app/apps/product-web/.next/standalone      ./
COPY  --chown=shop:nodejs --from=installer /app/node_modules/sharp                        node_modules/sharp

# Use /tmp/next/cache for storeing all caching
RUN mkdir -p /tmp/next/cache \
  && chown shop:nodejs /tmp/next/cache \
  && ln -s /tmp/next/cache /app/apps/product-web/.next/cache

USER shop

EXPOSE $PORT

ENV HOSTNAME 0.0.0.0

# I have a custom script to start nodejs
RUN chmod +x scripts/startup-web.sh

ENTRYPOINT ["sh", "-c", "scripts/startup-web.sh"]
```

**Comment by alesacchi** [comment]
Hi, I'm getting this error using nextjs 14.2.29

```
⨯ Failed to write image to cache XVcVL+UMa7JaZSZKj8v8lrqXCEVrGiDxqfQa9Ukzd10= [Error: ENOENT: no such file or directory, mkdir '/app/.next/cache'] {
  errno: -2,
  code: 'ENOENT',
  syscall: 'mkdir',
  path: '/app/.next/cache'
}
```

did you get a workaround?

**Comment by kelvyn-thai** [comment]
If you're deploying your Next.js application on Kubernetes and hitting the same issue, you can reference the solutions I considered.
Root cause:
The container uses readOnlyRootFilesystem: true, but Next.js/Sharp needs a writable /app/.next/cache directory for runtime image optimization.
**Option 1 — Mount an emptyDir at /app/.next/cache**
Pros:
Minimal change / quick fix
Keeps readOnlyRootFilesystem: true
Keeps Next.js/Sharp image optimization
Cache can be regenerated when needed
Cons:
Cache is lost when the pod is recreated
Needs to be configured for each Next.js service (ideally through a shared Helm/platform config)
**Option 2 — Pre-optimize images and serve them from S3/CDN**
Pros:
No runtime filesystem/cache dependency
Simpler container runtime
Can support different image sizes/DPR if needed
Cons:
Requires an asset optimization/upload process
New assets need to be deployed/uploaded across environments unless automated
**Option 3 — Skip Next.js image optimization**
In our case, we only serve a few static images, mainly to partners/staff on desktop. We can simply serve them directly without Next.js runtime optimization.
This avoids introducing additional K8s volumes or S3/CDN infrastructure. We lose automatic responsive image/DPR optimization, but the performance impact should be minimal for our use case.
For us, Option 3 is probably the simplest solution, while Option 1 is a good choice if runtime image optimization is still required.

<img width="764" height="611" alt="image" src="https://github.com/user-attachments/assets/f1231591-4676-4f83-b953-b5bc4adb8654" />
